src/d3a/models/strategy/storage.py

```diff
@@ -285,6 +287,16 @@ def event_tick(self):
     def event_trade(self, *, market_id, trade):
         market = self.area.get_future_market_from_id(market_id)
         super().event_trade(market_id=market_id, trade=trade)
+
+        if isinstance(trade.offer, Offer) and trade.offer.seller == self.owner.name:
+            offer = [o for o in self.offers.sold[market_id] if o.id == trade.offer.id][0]
+            assert trade.offer.price / trade.offer.energy >= \
+                offer.price / offer.energy - FLOATING_POINT_TOLERANCE
+        elif isinstance(trade.offer, Bid) and trade.offer.buyer == self.owner.name:
+            bid = [b for b in self.get_posted_bids(market) if b.id == trade.offer.id][0]
+            assert trade.offer.price / trade.offer.energy <= \
+                bid.price/bid.energy + FLOATING_POINT_TOLERANCE
```

Comment:
Would be better to abstract these checks into separate methods under `class BaseStrategy` since they are reimplemented for PV, Load and Storage strategy.

Reply:
Yes, you're right! I was thinking the same and will change this.